### ISSUE TITLE
Sunlit 3.x uploads to external blog failing with "not found" immediately after successfully uploading photo

### ISSUE THREAD

**Issue by lildude** [microdotblog/sunlit#130]
I'm writing my own [micro blogging interface](https://github.com/lildude/micropub-github-pages) to post directly to a GitHub repo (still 🚧) hosted on Heroku. Prior to updating to Sunlit 3, image uploads worked and Sunlit successfully uploaded the image to the media endpoint and then in a subsequent `POST`, the content. Both requests arrived and succeeded.

I updated to Sunlit 3 and now posting fails immediately after uploading the photo with "not found":

![IMG_5613](https://user-images.githubusercontent.com/627280/95880026-f3a25d80-0d6e-11eb-9216-43941b9463a3.png)

I've not changed anything on my media endpoint and it is still returning a 201 status code with the image URL in the `Location` header.

Uploading photos to the same endpoint using the micro.blog iOS app works as expected and I can see both the `POST` uploading the image to the media endpoint and then the subsequent `POST` for the content itself.

The same behaviour is seen with the Sunlit 3.2 beta (used for the screenshot).

I've tried signing out and in again and then re-adding my external blog and the behaviour doesn't change.

Is Sunlit now expecting the returned image URL to be instantly accessible or is something else at play?

**Comment by cheesemaker** [comment]
Can you try the latest 3.2 beta and see if it still fails? We had an issue with 3.1 that affected Micropub support.

**Comment by lildude** [comment]
@cheesemaker Already done that. 

> The same behaviour is seen with the Sunlit 3.2 beta (used for the screenshot).

I downloaded it and tested again about 15 mins before posting this issue.

**Comment by lildude** [comment]
Oh wait. Just seen build 379 is now available. Lemme try that.

**Comment by lildude** [comment]
Same problem with build 379 😢

**Comment by cheesemaker** [comment]
So you are seeing the images successfully POST and the HTML successfully POSTing as well?

**Comment by lildude** [comment]
No. Only the image POST. The HTML POST never happens. The "not found" error occurs immediately after the image POST.

Here's an excerpt from my heroku logs including some of my own debugging:

```
2020-10-13T16:15:15.213340+00:00 app[web.1]: I, [2020-10-13T16:15:15.213161 #4]  INFO -- : {"file"=>{"filename"=>"image.jpg", "type"=>"image/jpeg", "name"=>"file", "tempfile"=>#<Tempfile:/tmp/RackMultipart20201013-4-1rsj3bd.jpg>, "head"=>"Content-Disposition: form-data; name=\"file\"; filename=\"image.jpg\"\r\nContent-Type: image/jpeg\r\n"}, "site"=>"dev"}

2020-10-13T16:15:18.271051+00:00 heroku[router]: at=info method=POST path="/micropub/dev/media" host=micropub2github.herokuapp.com request_id=7a4aba35-0dc6-4ea3-b502-42d57639f4cc fwd="150.143.248.26" dyno=web.1 connect=0ms service=19016ms status=201 bytes=269 protocol=https

2020-10-13T16:15:18.271204+00:00 app[web.1]: 150.143.248.26 - - [13/Oct/2020:16:15:18 +0000] "POST /micropub/dev/media HTTP/1.1" 201 - 3.0980
```

If I post the exact same image from the micro.blog iOS app, I see:

```
2020-10-13T16:16:31.547549+00:00 app[web.1]: I, [2020-10-13T16:16:31.547376 #4]  INFO -- : {"file"=>{"filename"=>"image.jpg", "type"=>"image/jpeg", "name"=>"file", "tempfile"=>#<Tempfile:/tmp/RackMultipart20201013-4-1kb68cw.jpg>, "head"=>"Content-Disposition: form-data; name=\"file\"; filename=\"image.jpg\"\r\nContent-Type: image/jpeg\r\n"}, "site"=>"dev"}

2020-10-13T16:16:33.928129+00:00 heroku[router]: at=info method=POST path="/micropub/dev/media" host=micropub2github.herokuapp.com request_id=26100ec5-59f5-425a-89fb-957c348a71eb fwd="150.143.248.26" dyno=web.1 connect=1ms service=3912ms status=201 bytes=269 protocol=https

2020-10-13T16:16:33.926875+00:00 app[web.1]: 150.143.248.26 - - [13/Oct/2020:16:16:33 +0000] "POST /micropub/dev/media HTTP/1.1" 201 - 2.4295

2020-10-13T16:16:34.225501+00:00 app[web.1]: I, [2020-10-13T16:16:34.225360 #4]  INFO -- : {:content=>"Testing", :h=>"entry", :photo=>[["https://lildude.github.io/dev-micropub-pages/images/image.jpg"]], :"mp-photo-alt"=>[""], :name=>"", :type=>:note, :published=>"2020-10-13 16:16:34 +0000"}

2020-10-13T16:16:38.449281+00:00 heroku[router]: at=info method=POST path="/micropub/dev" host=micropub2github.herokuapp.com request_id=bcf21888-1daf-4b35-9cbe-8030efaf69aa fwd="150.143.248.26" dyno=web.1 connect=1ms service=4251ms status=201 bytes=260 protocol=https

2020-10-13T16:16:38.448127+00:00 app[web.1]: 150.143.248.26 - - [13/Oct/2020:16:16:38 +0000] "POST /micropub/dev HTTP/1.1" 201 - 4.2485
```

**Comment by cheesemaker** [comment]
So that error is from a 404 error. I can't tell from your logs if the image upload is succeeding or not. If it is, then it's the attempt to upload the HTML that's getting the 404. This is probably a dumb question, but have you tried deleting and re-installing Sunlit?

**Comment by lildude** [comment]
The uploads do succeed and the response is those lines with `201` which is the status code confirming success. I have also reinstalled the app. If this 404 is from the HTML post, it's not going to my endpoint as the request never appears in the logs. Is there any way I can enable debugging in the app and view the URL producing the 404?

**Comment by lildude** [comment]
I've reinstalled the beta again and now I can't even re-add my external blog. I can't sign out either and force quitting doesn't help. 🤔 Seems like something is being cached somewhere, possibly caching something that is causing the "not found" error too.

**Comment by cheesemaker** [comment]
One of the new features in 3.2 that we are testing out is that Sunlit now shares a login with Micro.blog. So if you wanted to sign out of Micro.blog, you'd need to sign out both in Sunlit and in Micro.blog. 

What happens when you try to re-add your external blog? My guess is that Sunlit somehow isn't finding the correct endpoint in the initial Micropub discovery. Btw, if you can provide a test account, I'm happy to take a look at this in the debugger to try and see what's going on.

**Comment by lildude** [comment]
> One of the new features in 3.2 that we are testing out is that Sunlit now shares a login with Micro.blog. So if you wanted to sign out of Micro.blog, you'd need to sign out both in Sunlit and in Micro.blog.

Oh wow! So I need to sign out of _both_ the Sunlit iOS app _and_ the Micro.blog app!? That is so unexpected and unintuitive. I'd expect the "Sign out of Micro.blog" link in Sunlit to do that _just_ in the Sunlit app, independent of the Micro.blog iOS app. That said, it's done the trick and I've now managed to sign out of Sunlit.

> What happens when you try to re-add your external blog? My guess is that Sunlit somehow isn't finding the correct endpoint in the initial Micropub discovery.

I still can't re-add back my external blog. I go through the process as before but it never appears in the list of blogs. I've worked out why...

My auth endpoint is rejecting the final auth request because there is no `me` parameter which looks to be the same issue at https://github.com/microdotblog/sunlit/issues/124. This is using 3.2 (379).

> Btw, if you can provide a test account, I'm happy to take a look at this in the debugger to try and see what's going on.

I can't really do this as there really isn't an account system behind it. My auth endpoint is an modified fork of https://github.com/barryf/acquiescence which auths against GitHub using a hardcoded username and the micropub endpoint uses the GitHub API to update the content in a configured repo.

I have managed to get the repo built and the simulator running locally so can troubleshoot from that side of things if it helps. 

It's also worth pointing out I've managed to reproduce the inability to add my external blog using the 3.2 branch locally in the simulator.

**Comment by manton** [comment]
@lildude The latest IndieAuth spec has removed `me` except in the initial auth request, so we removed it from Sunlit in these betas. For compatibility with more servers we may need to add it back.

**Comment by cheesemaker** [comment]
Do we need to add `me` back in?

**Comment by cheesemaker** [comment]
I believe that this is now a duplicate of #124 
I have added back the `me` parameter and there is a new beta build 3.2 (build 381) with this in it. Please check this and see if it is resolved. Thanks!

**Comment by lildude** [comment]
Will do as soon as TestFlight shows it for me. Still showing 379 here in 🇬🇧.

**Comment by paulrobertlloyd** [comment]
@manton Beyond the spec, can you point to a discussion where this change was mentioned and the reasons for it? I’m not sure my server distinguishes between initial and subsequent authentication requests. I find IndieAuth and the authentication flow pretty confusing to be hones, especially given there are so many different components at play (clients, servers, third-parties). I want to be spec compliant as far as possible, but this change confuses me even further!

(As to Sunlit, yes please keep using the older authentication method!)

**Comment by lildude** [comment]
I've tested with build 381 and we're all good now. I can re-add my external blog and posts are now heading to my external blog again, though not to my media endpoint like the micro.blog app does, but it does highlight I broke something recently 😊. 

So this does indeed appear to be a duplicate of https://github.com/microdotblog/sunlit/issues/124 and thus caused by the lack of `me` parameter.

I'm happy for this to be closed when everyone else is.
